pkg/detector/preemption.go

```diff
@@ -227,3 +228,48 @@ func (d *ResourceDetector) fetchResourceTemplate(rs policyv1alpha1.ResourceSelec
 	}
 	return resourceTemplate, nil
 }
+
+// HandleDeprioritizedPropagationPolicy responses to priority change of a PropagationPolicy,
+// if the change is from high priority (e.g. 5) to low priority(e.g. 3), it will
+// check if there is another PropagationPolicy could preempt the targeted resource,
+// and put the PropagationPolicy in the queue to trigger preemption.
+func (d *ResourceDetector) HandleDeprioritizedPropagationPolicy(oldPolicy policyv1alpha1.PropagationPolicy, newPolicy policyv1alpha1.PropagationPolicy) {
+	if (newPolicy.Spec.Priority != nil && oldPolicy.Spec.Priority != nil) && *newPolicy.Spec.Priority < *oldPolicy.Spec.Priority {
+		klog.Infof("PropagationPolicy(%s/%s) priority changed from %d to %d", newPolicy.GetNamespace(), newPolicy.GetName(), *oldPolicy.Spec.Priority, *newPolicy.Spec.Priority)
+		policies, err := d.propagationPolicyLister.ByNamespace(newPolicy.GetNamespace()).List(labels.Everything())
+		if err != nil {
+			klog.Errorf("Failed to list PropagationPolicy from namespace: %s, error: %v", newPolicy.GetNamespace(), err)
+			return
+		}
+
+		// TODO(@RainbowMango): Should sort the listed policies to ensure the
+		// higher priority PropagationPolicy be process first to avoid possible
+		// multiple preemption.
```

Comment:
I think it's necessary. in the end, only the policy with the highest priority and preemption-enabled can preempt it, and we don’t need to re-enqueue the rest of the policies.

Reply:
Not exactly, since we don't know if the potential policy matches the target resource, so we have to en-queue all of them.

These comments want to address the case that there is more than one policy that could perform the preemption, but with different priorities.  We'd better ensure en-queue the high-priority policy first, then the low-priority policy won't perform the preemption. Otherwise, the high-priority policy may preempt the low-priority policy again.(That's what I mean multiple preemption).

Note that, en-queue high-priority policy can't guarantee high-priority policy could perform preemption ahead of the low-priority one. It depends on the speed of the reconciliation routine. That's just a relief.

Reply:
I leave the TODO here because we don't have convenient helper functions to sort the policy yet, i think we can get back on the TODO task after that.

Reply:
> I leave the TODO here because we don't have convenient helper functions to sort the policy yet, i think we can get back on the TODO task after that.

+1
> Not exactly, since we don't know if the potential policy matches the target resource, so we have to en-queue all of them.

we can resolve this problem by checking if the intersection of their `ResourceSelector` exists.

Reply:
Yes we can, but I'm hesitant to do it for this `temporary` solution. I hope we can keep it simple.

Reply:
Given that it's a temporary solution, I totally agree.